### PR TITLE
docs: add redirect and update updating-to-v11 to v12

### PR DESCRIPTION
**Pull request by kapunahelewong**
This PR changes the `updating-to-version-11.md ` to `updating-to-version-12.md`, adds v12 info, and adds a redirect for the new file.

### CODE REVIEW

File: aio/content/guide/updating-to-version-12.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+* Add min and max Forms validators.
+  See [PR #39063](https://github.com/angular/angular/pull/39063).
+* Add `emitEvent` option for `AbstractControl` class methods.
+  See [PR #31031](https://github.com/angular/angular/pull/31031).
+* Support APP_INITIALIZER to work with observables.
+  See [PR #31031](https://github.com/angular/angular/pull/31031).
+* `HttpClient` supports specifying request metadata.
+  See [PR #25751](https://github.com/angular/angular/pull/25751).
+* Linked library code will no longer contain any legacy `i18n` message ids.
+For this reason, libraries that existed before version 12 that contained `i18n` template tags need to inform their clients that the message ids will have changed and instruct them on how to migrate them.
```
Comment (kapunahelewong) [inline aio/content/guide/updating-to-version-12.md:63]: @petebacondarwin what do you think of having this note you supplied in https://github.com/angular/angular/pull/41485#issuecomment-818745194 in this location?

Comment (petebacondarwin) [reply aio/content/guide/updating-to-version-12.md:63]: Generally this appears to be the right place to put this message.

I think we should have a consistent way of talking about these concepts. In this doc we have `ng-linker` and "linked library", but we have used the phrase "partial" compilation, e.g. in the angular compiler options.

I would put the two bullet points that mention the new library format next to each other. Can we have sub-bullets? Perhaps it could be worded something like:

```
* Support for publishing libraries with Ivy:
  - Libraries can now be compiled in "partial" compilation mode to generate ivy compatible output that will be "linked" when an application using that library is bundled.
  - New libraries created by the Angular CLI default to "partial" compilation mode, and do not support View Engine.
  - Libraries compiled in "partial" compilation mode will not contain legacy `i18n` message ids. If the library was previously compiled by View Engine, and contained legacy `i18n` message ids, then applications may have translation files that need to be migrated to the new message id format.
```

We should also document, somewhere, how to do this migration (@crisbeto ?) and then link to that from here.

Comment (crisbeto) [reply aio/content/guide/updating-to-version-12.md:63]: I talked to @aikidave about documenting the migration a few weeks ago, but I don't know how far along it is.

File: aio/content/guide/updating-to-version-12.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+
+<!-- Do we have new items to add here? -->
+
+{@a deprecations}
+### New Deprecations
+
+| Area                          | API or Feature                                     | May be removed in |
+| ----------------------------- | -------------------------------------------------- | ----------------- |
+| `@angular/core/testing`       | Rename `async` to `waitForAsync`                       | <!--v11--> v13 |
```
Comment (gkalpak) [inline aio/content/guide/updating-to-version-12.md:75]:
```suggestion
| `@angular/core/testing`       | Rename `async` to `waitForAsync`                   | <!--v11--> v13 |
```

File: aio/content/guide/updating-to-version-12.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+<!-- Removals for v12? -->
+{@a removals}
+### New Removals of Deprecated APIs
+
+The following APIs have been removed starting with version 11.0.0*:
+
+| Package          | API            | Replacement | Notes |
+| ---------------- | -------------- | ----------- | ----- |
+| `@angular/router`| `NavigationExtras#preserveQueryParams` | no action needed | NavigationExtras#preserveQueryParams has been removed from `@angular/router`.|
```
Comment (gkalpak) [inline aio/content/guide/updating-to-version-12.md:85]: Nit:

```suggestion
| `@angular/router` | `NavigationExtras#preserveQueryParams` | no action needed | `NavigationExtras#preserveQueryParams` has been removed from `@angular/router`.|
```

File: aio/content/guide/updating-to-version-12.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+| `@angular/core` | `CollectionChangeRecord` | no action needed | CollectionChangeRecord has been removed from `@angular/core`.|
```
Comment (gkalpak) [inline aio/content/guide/updating-to-version-12.md:86]:
```suggestion
| `@angular/core` | `CollectionChangeRecord` | no action needed | `CollectionChangeRecord` has been removed from `@angular/core`.|
```

File: aio/content/guide/updating-to-version-12.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+## Ivy features and compatibility
+
+Since version 9, Angular Ivy is the default rendering engine. If you haven't heard of Ivy, you can read more about it in the [Angular Ivy guide](guide/ivy).
+
+* Among other features, Ivy introduces more comprehensive type-checking within templates. For details, see [Template Type-checking](guide/template-typecheck).
+
+* For general guidance on debugging and a list of minor changes associated with Ivy, see the [Ivy compatibility guide](guide/ivy-compatibility).
+
+* For help with opting out of Ivy, see the instructions [here](guide/ivy#opting-out-of-angular-ivy).
```
Comment (gkalpak) [inline aio/content/guide/updating-to-version-12.md:99]: This might not be relevant for v12 apps (if it is not possible to build apps with `ViewEngine`).

File: aio/firebase.json [hunk 6]
```diff
@@ -32,7 +32,7 @@
       {"type": 301, "source": "/getting-started", "destination": "/start"},
       {"type": 301, "source": "/getting-started/:rest*", "destination": "/start/:rest*"},
       {"type": 301, "source": "/guide/displaying-data", "destination": "/start#template-syntax"},
-      {"type": 301, "source": "/guide/updating-to-version-10", "destination": "/guide/updating-to-version-11"},
```
Comment (gkalpak) [inline aio/firebase.json:35]: Is removing this redirect intentional? It means that links to `/guide/updating-to-version-10` will 404.

File: aio/content/guide/ivy-compatibility.md [hunk 7]
```diff
@@ -10,7 +10,7 @@ That said, some applications will likely need to apply some manual updates.
 
 If you're seeing errors, first temporarily [turn off Ivy](guide/ivy#opting-out-of-angular-ivy) in your `tsconfig.json` and re-start your app.
 
-If you're still seeing the errors, they are not specific to Ivy. In this case, you may want to consult the [general update guide](guide/updating-to-version-11). If you've opted into any of the new, stricter type-checking settings, you may also want to check out the [template type-checking guide](guide/template-typecheck).
+If you're still seeing the errors, they are not specific to Ivy. In this case, you may want to consult the [general update guide](guide/updating-to-version-12). If you've opted into any of the new, stricter type-checking settings, you may also want to check out the [template type-checking guide](guide/template-typecheck).
```
Comment (gkalpak) [inline aio/content/guide/ivy-compatibility.md:13]: This section (and possibly others in the guide) need updating, since turning off Ivy is not possible in v12 for applications (only for libraries) I think. (👈 confirmation needed)

File: aio/tests/deployment/shared/URLS_TO_REDIRECT.txt [hunk 8]
```diff
@@ -194,4 +194,4 @@
 /start/forms	/start/start-forms
 /start/routing	/start/start-routing
 /testing	/guide/testing
-/testing/first-app-tests.html	/guide/testing
\ No newline at end of file
+/testing/first-app-tests.html	/guide/testing
```
Comment (gkalpak) [inline aio/tests/deployment/shared/URLS_TO_REDIRECT.txt:197]: This newline causes the `test_aio` CI failure. Since this behavior is annoying, I submitted #41680.
(In the meantime, you can make CI pass by removing the newline.)

**Review by gkalpak — CHANGES_REQUESTED** [review]
@kapunahelewong: FYI, I've changed the PR target to `rc`, since it includes v12 specific changes, while `patch` still refers to v11.

File: aio/content/guide/updating-to-version-12.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+  See [Theming your Angular Material app](https://material.angular.io/guide/theming).
+* CLI production-ready webpack 5 support.
+  Remove usage and support of Webpack 4.
+  <!-- Noting specifically here that these are breaking in case all changes aren't breaking. -->
+  BREAKING CHANGE: Webpack 5 lazy loaded file name changes.
+  BREAKING CHANGE: Webpack 5 web worker support.
+  See [PR #20466](https://github.com/angular/angular-cli/pull/20466).
+* Add build option `inlineStyleLanguage` that allows a project to define the stylesheet language used in an application's inline component styles.
+  See [PR #20514](https://github.com/angular/angular-cli/pull/20514).
+* BREAKING CHANGE: Enable critical CSS inlining by default.
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md:43]: Yes, so those are notes to me in that I understand that those are certainly breaking, my question is, are all the changes tentatively in this section breaking ([see line 20 comment](https://github.com/angular/angular/pull/41625/files#diff-f332eb8445f10de5c8ff582d53dad312a95a294eeede76444c43b04eff26214aR20))? 

If some are breaking and some aren't, then we should either specify which in the list are breaking (and adjust the header accordingly) or have the breaking ones in their own section. For example, the `inlineStyleLanguage` change above this one seems that it wouldn't be breaking, but I need confirmation. Of the list items that don't specify breaking, would you know which are breaking and which aren't, @gkalpak? If you know specific ones, I'll take that too!

File: aio/content/guide/updating-to-version-12.md [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+</div>
+
+{@a breaking-changes}
+### New Breaking changes in Angular version 12
+
+<!-- Are all these breaking? If all are not, should we separate them into breaking and non-breaking?-->
+* Deprecate support for IE 11.
+  See [Deprecated APIs and features](guide/deprecations) and [Microsoft 365 apps say farewell to Internet Explorer 11 and Windows 10 sunsets Microsoft Edge Legacy](https://techcommunity.microsoft.com/t5/microsoft-365-blog/microsoft-365-apps-say-farewell-to-internet-explorer-11-and/ba-p/1591666).
+* Deprecate `ViewEngine`.
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md:23]: Thanks so much for this, @gkalpak!

Hey @aikidave, can you point me in the right direction for confirming this?

File: aio/content/guide/updating-to-version-12.md (hunk 2 above)
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md:75]: Hey @gkalpak, I'm not seeing the diff here. Also, does this need to be removed entirely from the doc as this is from the current doc? https://angular.io/guide/updating-to-version-11#new-deprecations

File: aio/content/guide/updating-to-version-12.md (hunk 3 above)
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md:85]: This is also from the current doc. Does this info stay? Are there cases where a section is here for multiple releases?

File: aio/content/guide/updating-to-version-12.md (hunk 4 above)
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md:86]: Same question as above:

> This is also from the current doc. Does this info stay? Are there cases where a section is here for multiple releases?

File: aio/content/guide/updating-to-version-12.md (hunk 5 above)
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md:99]: Same question as above:

> This is also from the current doc. Does this info stay? Are there cases where a section is here for multiple releases?

File: aio/content/guide/ivy-compatibility.md (hunk 7 above)
Comment (kapunahelewong) [reply aio/content/guide/ivy-compatibility.md:13]: Thanks for highlighting this, @gkalpak.

Hey @alxhub could you shed some light on this for us?

File: aio/content/guide/updating-to-version-12.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply aio/content/guide/updating-to-version-12.md:43]: I see. I can't be 100% sure from the descriptions, but there seems like there are a couple non-breaking changes there:
- ``Add build option `inlineStyleLanguage` ``
- `Add min and max Forms validators`
- `` `HttpClient` supports specifying request metadata``
- `Support for publishing libraries with Ivy`

Better to confirm with someone :grin:
If you have commits corresponding to each change, you can (theoretically) easily check whether the change is a breaking one by looking for a `BREAKING CHANGE` section in the commit message.

File: aio/content/guide/updating-to-version-12.md (hunk 2 above)
Comment (gkalpak) [reply aio/content/guide/updating-to-version-12.md:75]: The difference is that there is some extra trailing space in the second column.

If this was present in the v11 doc, though, then it is not a new deprecation in v12, so it should be removed I guess.

File: aio/content/guide/updating-to-version-12.md (hunk 3 above)
Comment (gkalpak) [reply aio/content/guide/updating-to-version-12.md:85]: AFAICT, this docs should only contain deprecations/removals/breaking changes pertaining to the v12 release. So, if something was included in the v11 doc, it should probably not be in the v12 doc as well :grin:

File: aio/content/guide/updating-to-version-12.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md:42]: Oh thank you, @gkalpak! I didn't know about the `BREAKING CHANGE` section in the commit message. I'll check them. 🌸

**Comment by jelbourn** [comment]
FYI I met w/ @kapunahelewong to go over the context for the changes here to clarify what's breaking/non-breaking

File: aio/content/guide/updating-to-version-12.md (hunk 4 above)
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md]: resolved offline

File: aio/content/guide/updating-to-version-12.md (hunk 5 above)
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md]: resolved offline

File: aio/content/guide/updating-to-version-12.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md]: resolved offline

File: aio/content/guide/ivy-compatibility.md (hunk 7 above)
Comment (kapunahelewong) [reply aio/content/guide/ivy-compatibility.md:13]: resolved offline

**Review by TeriGlover — APPROVED** [review]
Editing review; a few suggestions.

File: aio/ngsw-config.json [hunk 11]
```diff
@@ -124,9 +124,9 @@
     "!/guide/service-worker-configref",
     "!/guide/service-worker-configref.html",
     "!/guide/service-worker-configref/",
-    "!/guide/updating-to-version-10",
-    "!/guide/updating-to-version-10.html",
-    "!/guide/updating-to-version-10/",
```
Comment (gkalpak) [inline aio/ngsw-config.json:129]: I don't think we should be removing these (since we still want `*-10` t0 be redirected.
I believe you could change the patterns to the following, so we don't have to update them every time:

```json
"!/guide/updating-to-version-\\d{1,}",
"!/guide/updating-to-version-\\d{1,}.html",
"!/guide/updating-to-version-\\d{1,}/",
```

File: aio/content/guide/ivy-compatibility.md (hunk 7 above)
Comment (gkalpak) [reply aio/content/guide/ivy-compatibility.md:13]: What was the resolution?

File: aio/content/guide/updating-to-version-12.md (hunk 5 above)
Comment (gkalpak) [reply aio/content/guide/updating-to-version-12.md]: What was the resolution?

File: aio/content/errors/NG6999.md [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+@shortDescription Invalid @NgModule() metadata
+
+@description
+This error represents the import or export of an `@NgModule()` that doesn't have valid metadata.
+
+@debugging
+The library might have been processed with `ngcc`.
+If this is the case, try removing and reinstalling `node_modules`.
+If the error persists, the issue might be due to the library being published for Angular Ivy and Angular version 12 and above.
+Check the peer dependencies to ensure that you're using a compatible version of Angular.
```
Comment (gkalpak) [inline aio/content/errors/NG6999.md:12]: Does this file really belong to this PR? It seems unrelated 😕

File: aio/content/guide/updating-to-version-12.md (hunk 5 above)
Comment (gkalpak) [reply aio/content/guide/updating-to-version-12.md]: NVM, I see this section was removed from the doc :+1:

File: aio/content/guide/ivy-compatibility.md (hunk 7 above)
Comment (gkalpak) [reply aio/content/guide/ivy-compatibility.md:13]: NVM, I see this section was removed from the doc 👍

File: aio/content/guide/updating-to-version-12.md [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+* There is a new Angular Material Sass API for `@use`.
+  Run the migration script `ng g @angular/material:themingApi` to switch all your Sass imports for Angular CDK and Angular Material to the new API and `@use`.
+  Sass imports from `@angular/material/theming` are deprecated.
+* Support for publishing libraries with Ivy:
+  - You can now compile libraries in [_partial_ compilation mode](https://angular.io/guide/angular-compiler-options#compilationmode) to generate ivy compatible output that will be _linked_ when an application using that library is bundled.
+  - New libraries you create with the Angular CLI default to partial compilation mode, and do not support View Engine.
+  - Libraries compiled in partial compilation mode will not contain legacy `i18n` message ids.
+  If the library was previously compiled by View Engine, and contained legacy `i18n` message ids, then applications may have translation files that you'll need to migrate to the new message id format.
+  <!-- For more information see, [](NEED LINK). -->
```
Comment (gkalpak) [inline aio/content/guide/updating-to-version-12.md:96]: Can this be removed now?

```suggestion
```

File: aio/content/guide/updating-to-version-12.md [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+  See [Angular CLI PR #20128](https://github.com/angular/angular-cli/pull/20128).
+* Previously, the Forms module ignored `min` and `max` attributes defined on the `<input type="number">`
+  Now these attributes trigger `min` and `max` validation logic in cases where `formControl`, `formControlName` or `@ngModel()` directives are also present on a given input.
+  See [PR #39063](https://github.com/angular/angular/pull/39063).
+
+
+{@a deprecations}
+### New Deprecations
+
+* Deprecate support for IE 11.
```
Comment (gkalpak) [inline aio/content/guide/updating-to-version-12.md:86]:
```suggestion
* Support for IE 11 has been deprecated.
```

File: aio/content/guide/updating-to-version-12.md [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+  See [Deprecated APIs and features](guide/deprecations) and [Microsoft 365 apps say farewell to Internet Explorer 11 and Windows 10 sunsets Microsoft Edge Legacy](https://techcommunity.microsoft.com/t5/microsoft-365-blog/microsoft-365-apps-say-farewell-to-internet-explorer-11-and/ba-p/1591666).
+* There is a new Angular Material Sass API for `@use`.
+  Run the migration script `ng g @angular/material:themingApi` to switch all your Sass imports for Angular CDK and Angular Material to the new API and `@use`.
+  Sass imports from `@angular/material/theming` are deprecated.
```
Comment (gkalpak) [inline aio/content/guide/updating-to-version-12.md:90]: I would move this sentence to the beginning of the bullet-point to make clear what is being deprecated.

File: aio/content/guide/updating-to-version-12.md [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+* Support for publishing libraries with Ivy:
```
Comment (gkalpak) [inline aio/content/guide/updating-to-version-12.md:91]: This sounds like publishing libraries with Ivy is deprecated. How about:

```suggestion
* Support for publishing libraries with View Engine has been deprecated:
```

File: aio/content/guide/updating-to-version-12.md [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+  - You can now compile libraries in [_partial_ compilation mode](https://angular.io/guide/angular-compiler-options#compilationmode) to generate ivy compatible output that will be _linked_ when an application using that library is bundled.
+  - New libraries you create with the Angular CLI default to partial compilation mode, and do not support View Engine.
+  - Libraries compiled in partial compilation mode will not contain legacy `i18n` message ids.
+  If the library was previously compiled by View Engine, and contained legacy `i18n` message ids, then applications may have translation files that you'll need to migrate to the new message id format.
+  <!-- For more information see, [](NEED LINK). -->
+  - For context, see [Issue #38366](https://github.com/angular/angular/issues/38366).
```
Comment (gkalpak) [inline aio/content/guide/updating-to-version-12.md:97]: I think it would be helpful to mention that one can still build a library with View Engine and link to some place (where?) that contains info on how to do it?

File: aio/content/errors/NG6999.md [hunk 18]
```diff
@@ -0,0 +1,12 @@
+@name Invalid metadata
```
Comment (gkalpak) [inline aio/content/errors/NG6999.md:1]: Does this file really belong to this PR? It seems unrelated 😕

File: aio/ngsw-config.json (hunk 11 above)
Comment (gkalpak) [inline aio/ngsw-config.json:129]: I don't think we should be removing these (since we still want `*-10` t0 be redirected.
I believe you could change the patterns to the following, so we don't have to update them every time:

```json
"!/guide/updating-to-version-\\d{1,}",
"!/guide/updating-to-version-\\d{1,}.html",
"!/guide/updating-to-version-\\d{1,}/",
```

**Review by gkalpak — CHANGES_REQUESTED** [review]

File: aio/content/errors/NG6999.md (hunk 18 above)
Comment (kapunahelewong) [reply aio/content/errors/NG6999.md:1]: 🧐 awwww I have no idea how I managed that. Goodness! Thank you for catching this! 😁

File: aio/content/guide/updating-to-version-12.md [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+### Breaking changes in Angular version 12
+
+* Add support for TypeScript 4.2.
+  TypeScript 4 is no longer supported.
+  The supported range of TypeScript versions is 4.2.3 to 4.2.x.
+  See [PR #41158](https://github.com/angular/angular/pull/41158).
+* Angular CDK and Angular Material internally now use the new [Sass module system](https://sass-lang.com/blog/the-module-system-is-launched), which is actively maintained by the Sass team at Google.
+  Consequently, applications can no longer consume Angular CDK/Material's Sass with the `node-sass` npm package.
+  `node-sass` is unmaintained and does not support newer Sass features. Instead, applications must use the `sass` npm package, or the `sass-embedded` npm package for the `sass-embedded` beta.
+* The Angular tooling now uses Webpack 5 to build applications. Webpack 4 usage and support has been removed.
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md:58]: Here, the word tooling is a noun, so it's what uses Webpack.

File: aio/content/guide/updating-to-version-12.md [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+{@a breaking-changes}
+### Breaking changes in Angular version 12
+
+* Add support for TypeScript 4.2.
+  TypeScript 4 is no longer supported.
+  The supported range of TypeScript versions is 4.2.3 to 4.2.x.
+  See [PR #41158](https://github.com/angular/angular/pull/41158).
+* Angular CDK and Angular Material internally now use the new [Sass module system](https://sass-lang.com/blog/the-module-system-is-launched), which is actively maintained by the Sass team at Google.
+  Consequently, applications can no longer consume Angular CDK/Material's Sass with the `node-sass` npm package.
+  `node-sass` is unmaintained and does not support newer Sass features. Instead, applications must use the `sass` npm package, or the `sass-embedded` npm package for the `sass-embedded` beta.
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md:57]: I added the links (thank you!!) but didn't find anything about what "`sass-embedded` beta" means. I found [this repo](https://github.com/sass/embedded-protocol) but is that what you're thinking? I also found [this tweet](https://twitter.com/SassCSS/status/1383151350851072003), but nothing like an article or blog post.

File: aio/content/guide/updating-to-version-12.md [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+* The Angular tooling now uses Webpack 5 to build applications. Webpack 4 usage and support has been removed.
+  You don't need to make any project level configuration changes to use the upgraded Webpack version when using the official Angular builders.
+  Custom builders based on this package that use the experimental programmatic APIs may need to be updated to become compatible with Webpack 5.
+  See [PR #20466](https://github.com/angular/angular-cli/pull/20466).
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md:62]: I can't reproduce this formatting issue. Is your screen zoomed in by chance? I've broken the layout by forgetting I zoomed in.

File: aio/content/guide/updating-to-version-12.md [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+* Webpack 5 generates similar but differently named files for lazy-loaded JavaScript files in development configurations when the `namedChunks` option is enabled.
+  For the majority of users this change should have no effect on the application or build process.
+  Production builds should also not be affected as the `namedChunks` option is disabled by default in production configurations.
+  However, if a project's post-build process makes assumptions as to the file names then adjustments may need to be made to account for the new naming paradigm.
+  Such post-build processes could include custom file transformations after the build, integration into service-side frameworks, or deployment procedures.
+  An example of a development file name change is `lazy-lazy-module.js` becoming `src_app_lazy_lazy_module_ts.js`.
+  See [PR #20466](https://github.com/angular/angular-cli/pull/20466).
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md:69]: Same as above: https://github.com/angular/angular/pull/41625#discussion_r620659820

File: aio/content/guide/updating-to-version-12.md [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+* Webpack 5 now includes web worker support.
+  However, the structure of the URL within the worker constructor must be in a specific format that differs from the current requirement.
+  To update Web worker usage, where `./app.worker` is the actual worker name, change `new Worker('./app.worker', ...)` to `new Worker(new URL('./app.worker', import.meta.url), ...)`.
+  See [PR #20466](https://github.com/angular/angular-cli/pull/20466).
+* Critical CSS inlining is now enabled by default.
+  To turn this off, set `inlineCritical` to false.
+  See [Angular CLI PR #20096](https://github.com/angular/angular-cli/pull/20096) and the [Style preprocessor options](https://angular.io/guide/workspace-config#optimization-configuration) section of [Angular workspace configuration](https://angular.io/guide/workspace-config).
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md]: Same as above: https://github.com/angular/angular/pull/41625#discussion_r620659820

Leaving open for reference.

File: aio/content/guide/updating-to-version-12.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply aio/content/guide/updating-to-version-12.md]: I don't know much about it either, so feel free to leave it as is :grin:

File: aio/firebase.json (hunk 6 above)
Comment (gkalpak) [inline aio/firebase.json:35]: Is removing this redirect intentional? It means that links to `/guide/updating-to-version-10` will 404.
(Also, tests need updating if you do want to remove this redirect.)

**Review by gkalpak — APPROVED** [review]
A couple of minor comments. Otherwise lgtm (as soon as the CI failure is fixed) :+1:

File: aio/content/guide/updating-to-version-12.md [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+
+* Support for Internet Explorer 11 has been deprecated.
+  See [Deprecated APIs and features](guide/deprecations) and [Microsoft 365 apps say farewell to Internet Explorer 11 and Windows 10 sunsets Microsoft Edge Legacy](https://techcommunity.microsoft.com/t5/microsoft-365-blog/microsoft-365-apps-say-farewell-to-internet-explorer-11-and/ba-p/1591666).
+* Sass imports from `@angular/material/theming` are deprecated. There is a new Angular Material Sass API for `@use`.
+  Run the migration script `ng g @angular/material:themingApi` to switch all your Sass imports for Angular CDK and Angular Material to the new API and `@use`.
+* Support for publishing libraries with View Engine has been deprecated:
+  - You can now compile libraries in [_partial_ compilation mode](https://angular.io/guide/angular-compiler-options#compilationmode) to generate Ivy compatible output that will be _linked_ when an application using that library is bundled.
+  - New libraries you create with the Angular CLI default to partial compilation mode, and do not support View Engine. You can still build a library with View Engine. See [Creating libraries](guide/creating-libraries) for more information.
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md:93]: I have just such a thing cooking. May I interest you in https://github.com/angular/angular/pull/41485?  :stew:

File: aio/firebase.json (hunk 6 above)
Comment (kapunahelewong) [reply aio/firebase.json:35]: Ah ok, @gkalpak, I put it back. But since it redirects to 11 which in turn redirects, is that ok?

File: aio/content/guide/updating-to-version-12.md [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+This guide contains information related to updating to Angular version 12.
+
+## Updating CLI applications
+
+For step-by-step instructions on how to update to the latest Angular release and leverage our automated migration tools to do so, use the interactive update guide at [update.angular.io](https://update.angular.io).
+
+
+## Building applications with Ivy
+
+For libraries, View Engine will be deprecated and removed in version 13.
```
Comment (jelbourn) [inline aio/content/guide/updating-to-version-12.md:12]:
```suggestion
For libraries, View Engine is deprecated and will be removed in version 13.
```

File: aio/content/guide/updating-to-version-12.md [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+New libraries built with version 12 default to Ivy.
```
Comment (jelbourn) [inline aio/content/guide/updating-to-version-12.md:13]:
```suggestion
New libraries created with version 12 or later default to Ivy.
```

File: aio/content/guide/updating-to-version-12.md [hunk 27]
```diff
@@ -0,0 +1,103 @@
...
+
```
Comment (jelbourn) [inline aio/content/guide/updating-to-version-12.md:14]: I would include a link here to the documentation for `ng-linker` (when it's done) with something like
"More more information about distributing libraries with View Engine and Ivy, see ..."

File: aio/content/guide/updating-to-version-12.md [hunk 28]
```diff
@@ -0,0 +1,103 @@
...
+   For information about Angular's deprecation and removal practices, see [Angular Release Practices](guide/releases#deprecation-practices "Angular Release Practices: Deprecation practices").
+
+</div>
+
+* Applications can no longer build with View Engine by setting `enableIvy: false`.
+  Support for building libraries with View Engine, for backwards compatibility, is deprecated and will be removed in Angular version 13.
+  New libraries created with Angular version 12 will default to building and distributing with Ivy.
+  For more information, see [Creating Libraries](guide/creating-libraries).
+* The Ivy-based language service is now on by default.
```
Comment (jelbourn) [inline aio/content/guide/updating-to-version-12.md:27]:
```suggestion
* The Ivy-based IDE language service is now on by default.
```
I would probably also make "language service" a link here to more info

File: aio/content/guide/updating-to-version-12.md [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+  See [PR #1279](https://github.com/angular/vscode-ng-language-service/pull/1279).
+* The previous message ID generation algorithm is deprecated.
```
Comment (jelbourn) [inline aio/content/guide/updating-to-version-12.md:29]:
```suggestion
* Angular's View Engine-based algorithm for generating i18n message IDs is deprecated.
```
@petebacondarwin or @crisbeto is this proposed edit accurate? I want to cut out any ambiguity about which specific behavior is deprecated and how the new behavior is enabled.

File: aio/content/guide/updating-to-version-12.md [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+  Angular version 12 adds a new flag to `localize-extract` called `--migrateMapFile` which  generates a JSON file that can be used to map legacy message IDs to canonical ones.
+  There is also a new script called `localize-migrate` that can use the mapping file which `localize-extract` generates and migrate all of the IDs in the files that are passed in.
+  For better stability, if you are using Angular's `i18n`, run this migration to move to the new message ID generation algorithm.
+  If you don't run this migration, all your generated message IDs will  change when Angular removes the View Engine compiler.
```
Comment (jelbourn) [inline aio/content/guide/updating-to-version-12.md:33]:
```suggestion
  If you don't run this migration, all your generated message IDs will change when Angular removes the View Engine compiler.
```
extra space

File: aio/content/guide/updating-to-version-12.md [hunk 31]
```diff
@@ -0,0 +1,103 @@
...
+  See [PR #20128](https://github.com/angular/angular-cli/pull/20128).
+* Previously, the Forms module ignored `min` and `max` attributes defined on the `<input type="number">`.
+  Now these attributes trigger `min` and `max` validation logic in cases where `formControl`, `formControlName`, or `ngModel` directives are also present on a given input.
+  See [PR #39063](https://github.com/angular/angular/pull/39063).
+
+
+{@a deprecations}
+### New deprecations
+
+* Support for Internet Explorer 11 has been deprecated.
```
Comment (jelbourn) [inline aio/content/guide/updating-to-version-12.md:87]:
```suggestion
* Support for Internet Explorer 11 is deprecated.
```

File: aio/content/guide/updating-to-version-12.md (hunk 1 above)
Comment (aikidave) [reply aio/content/guide/updating-to-version-12.md]: Documented in https://github.com/angular/angular/pull/41828

**Review by aikithoughts — APPROVED** [review]
Reviewed-for: global-docs-approvers

File: aio/content/guide/updating-to-version-12.md (hunk 1 above)
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md]: Added link. Thanks, @aikidave.

**Comment by kapunahelewong** [comment]
Blocked by https://github.com/angular/angular/pull/41828, which must be merged first as updates in this PR link to the new Message IDs guide in https://github.com/angular/angular/pull/41828.

**Comment by kapunahelewong** [comment]
This PR shouldn't be merged before https://github.com/angular/angular/pull/41485

File: aio/content/guide/updating-to-version-12.md (hunk 29 above)
Comment (kapunahelewong) [reply aio/content/guide/updating-to-version-12.md:30]: I've added this in the interim as it is at least clearer than what I had.

File: aio/ngsw-config.json (hunk 11 above)
Comment (josephperrott) [reply aio/ngsw-config.json]: Unfortunately we need to update this as `\d{1,}` will match `10`,`11` **and** `12`.  Causing the `updating-to-version-12` to never actually be routable

**Review by josephperrott — APPROVED** [review]
LGTM
